DAGE-10: Added project structure

dantuzi opened the pull request:
### Notes
Adding the first project structure with logging and a simple dummy test

### Ticket
https://sease.atlassian.net/browse/DAGE-10

Commit:
DAGE-10: added project structure with basic files, loggin, entrypoint and dummy tests

dantuzi commented on rre-dataset-generator/README.md:
````diff
@@ -1,2 +1,104 @@
+First, create and activate a virtual environment using `uv`:
+
+```bash
+# Create the virtual environment in .venv
````
We should have a code block for each command so the README user can copy/paste the single command

nicolo-rinaldi commented on rre-dataset-generator/README.md:
````diff
@@ -1,2 +1,104 @@
+
+```bash
+# Create the virtual environment in .venv
+uv init
````
This should be for the first initialisation of the repo. If the uv configuration files (e.g. uv.lock, pyproject.toml) are already in the repo, only `uv sync` is needed.

nicolo-rinaldi commented on rre-dataset-generator/README.md:
````diff
@@ -1,2 +1,104 @@
+Execute the main script via the `typer` CLI, pointing to your configuration file:
+
+```bash
+python main.py run --arg1 --arg2 # example --config ./config.yaml --reasoning --save_reasoning
````
replace `python` with `uv run` to use the uv environment.

nicolo-rinaldi commented on rre-dataset-generator/README.md:
```diff
@@ -1,2 +1,104 @@
+uv pip install -e '.[dev,test]'
+
+# Activate the venv (Linux/macOS)
+source .venv/bin/activate
```
No need to install any dependency or interact with the environment, just `uv sync` is doing everything.

nicolo-rinaldi commented on rre-dataset-generator/README.md:
````diff
@@ -1,2 +1,104 @@
+
+Create a `config.yaml` file in the root directory. This file controls the entire generation process. E.g:
+
+```yaml
````
TO BE CHANGED

dantuzi commented on rre-dataset-generator/README.md:
```diff
@@ -1,2 +1,104 @@
+uv venv
+
+# Install all dependencies from pyproject.toml
+uv pip install -e '.[dev,test]'
```
It doesn't work to another computer

dantuzi replied:
Also, I see this tool `uv` creates a myproject.toml file with all dependencies to be downloaded.
@nicolo-rinaldi do you know where all those dependencies are found? I don't see the standard `requirements.txt` file

nicolo-rinaldi replied:
Yes, the dependencies are built automatically during the installation of the python packages. With uv, the installation must be done using `uv add` instead of `pip install`. In this way the file is written automatically.

dantuzi commented on rre-dataset-generator/dataset_generator/__init__.py:
I guess we don't need an empty file

nicolo-rinaldi replied:
`__init__.py` files are used in python to distinguish between normal folders and packages one. I don't know if it is needed, but in a lot of repo is like that.

dantuzi commented on rre-dataset-generator/README.md:
````diff
@@ -1,2 +1,104 @@
+outputExplanation: false # Whether the LLM should provide explanations
+```
+
+### 4. Running the Generator
````
It should be n.3

dantuzi commented on rre-dataset-generator/README.md:
```diff
@@ -1,2 +1,104 @@
+1.  Fetch documents from the specified search engine.
+2.  Generate or load queries.
+3.  Score the relevance for each (document, query) pair.
+4.  (Future) Save the output to the specified destination.
```
We need a section that describes how to run the unit tests

nicolo-rinaldi replied:
I'll write it down just to remind myself when updating the readme: you just need to run the following
```
pytest <folder-with-test-files>
```

dantuzi commented on rre-dataset-generator/tests/unit/test_dataset_generation_dummy.py:
```diff
@@ -0,0 +1,32 @@
+configure_logging()
+
+
+def test_argparse_yaml_with_list(caplog):
```
When you run the test, it fails with
```
ModuleNotFoundError: No module named 'dataset_generator.main'
```

nicolo-rinaldi replied:
solved

dantuzi commented on rre-dataset-generator/tests/unit/test_dataset_generation_dummy.py:
```diff
@@ -0,0 +1,32 @@
+        print(f"LOG: {record.message}")
+    
+    # Check if the log message appears in any of the log records
+    assert any("Starting dataset generation process..." in record.message 
```
What does it want to assert? Do we want to verify that all messages are "Starting dataset generation process..."?
BTW, at this stage, we simply want a dummy test because we want to validate that the tests are well configured

nicolo-rinaldi replied:
Changed to just dummy tests

dantuzi commented on rre-dataset-generator/tests/unit/test_dataset_generation_dummy.py:
```diff
@@ -0,0 +1,32 @@
+    # Check if the log message appears in any of the log records
+    assert any("Starting dataset generation process..." in record.message 
+              for record in caplog.records)
+    assert result.exit_code == 0
```
It's always preferable to leave a blank line at the end of each file

nicolo-rinaldi replied:
solved

dantuzi commented on rre-dataset-generator/dataset_generator/logger.py:
```diff
@@ -0,0 +1,114 @@
+    """
+    Formats log records as a single line of JSON.
+    
+    This formatter ensures that logs are machine-readable, which is ideal for
```
Why do we have logs formatted to be machine-readable? Our script is not supposed to be monitored by any of the log aggregator tools like ELK, Datadog, or Splunk. We want our logs to be human-readable

dantuzi commented on rre-dataset-generator/README.md:
```diff
@@ -1,2 +1,104 @@
+│   ├── logger.py                       # Structured logging configuration
+│   ├── models.py                       # Pydantic models for internal data structures
+│   ├── prompts.py                      # Prompts for the LLM 
+│   ├── llm/                            # Sub-package for Large Language Model services
```
A few folders don't exist

nicolo-rinaldi commented on rre-dataset-generator/dataset_generation.py:
I changed the `typer` framework with the `argparse` one. This is a choice mainly due to the fact that, in our framework, we want to pass the config "dictionary" after parsing to a different set of functions. I could not find a way to do it using `typer`, so I changed framework.

nicolo-rinaldi commented on rre-dataset-generator/dataset_generator/llm/.gitkeep:
Replaced with a `__init__.py` file, to show the IDE that the llm folder is a package.

nicolo-rinaldi commented on rre-dataset-generator/dataset_generator/search_engine/.gitkeep:
Replaced with a `__init__.py` file, to show the IDE that the llm folder is a package.

nicolo-rinaldi commented on rre-dataset-generator/Makefile:
Why do we have this file?

nicolo-rinaldi reviewed (commented):
# Reviewed
- all files but `logger.py`

nicolo-rinaldi committed:
# Changes
 - changed the folder  ->
 - modified  files into  to make both the folder visible to git and also let the IDE understand that that folder is a folder of a python package
 - reviewed file : left untouched
 - changed file  to deal with input parameter with the library  instead of
 Usage: typer [OPTIONS] [PATH_OR_MODULE] COMMAND [ARGS]...

 Run Typer scripts with completion, without having to create a package.

 You probably want to install completion for the typer command:
 $ typer --install-completion
 https://typer.tiangolo.com/

╭─ Arguments ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│   path_or_module      [PATH_OR_MODULE]  [default: None]                                                                                                                           │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --app                       TEXT  The typer app object/variable to use. [default: None]                                                                                           │
│ --func                      TEXT  The function to convert to Typer. [default: None]                                                                                               │
│ --version                         Print version and exit.                                                                                                                         │
│ --install-completion              Install completion for the current shell.                                                                                                       │
│ --show-completion                 Show completion for the current shell, to copy it or customize the installation.                                                                │
│ --help                            Show this message and exit.                                                                                                                     │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ utils   Extra utility commands for Typer apps.                                                                                                                                    │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯. This is just because I already worked with that library in another config file and it was hard to change my code.
 - created folders in  folder that where seen in the structure but not in the actual repo and added files , for the same reason said above
 - changed file in  to deal with  library instead of
 Usage: typer [OPTIONS] [PATH_OR_MODULE] COMMAND [ARGS]...

 Run Typer scripts with completion, without having to create a package.

 You probably want to install completion for the typer command:
 $ typer --install-completion
 https://typer.tiangolo.com/

╭─ Arguments ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│   path_or_module      [PATH_OR_MODULE]  [default: None]                                                                                                                           │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --app                       TEXT  The typer app object/variable to use. [default: None]                                                                                           │
│ --func                      TEXT  The function to convert to Typer. [default: None]                                                                                               │
│ --version                         Print version and exit.                                                                                                                         │
│ --install-completion              Install completion for the current shell.                                                                                                       │
│ --show-completion                 Show completion for the current shell, to copy it or customize the installation.                                                                │
│ --help                            Show this message and exit.                                                                                                                     │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ utils   Extra utility commands for Typer apps.                                                                                                                                    │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

dantuzi commented on rre-dataset-generator/README.md:
````diff
@@ -1,2 +1,109 @@
+
+## Project Structure
+
+```bash
````
Not a blocker, but I'm afraid this section will be really difficult to maintain.
I would propose to remove it entirely or keep only the main folders.
For example, instead of keeping the folder "search_engine" with all the files, we can just keep the folder "search_engine"

dantuzi commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -0,0 +1,29 @@
+def parse_args():
+    parser = argparse.ArgumentParser(description='Parse arguments for dataset_generator CLI.')
+
+    parser.add_argument('-d', '--dummy', type=str,
```
I guess we don't need it

nicolo-rinaldi replied:
This is a temporary file to do a dummy test.

dantuzi commented on rre-dataset-generator/pyproject.toml:
```diff
@@ -0,0 +1,16 @@
+]
+
+[project.scripts]
+dataset-generator = "main:app"
```
I saw that when I run the `uv sync` command with this configuration, it raises a Warning saying
```
warning: Skipping installation of entry points (`project.scripts`) because this project is not packaged; to install entry points, set `tool.uv.package = true` or define a `build-system`
```

dantuzi commented on rre-dataset-generator/src/logger.py:
```diff
@@ -0,0 +1,114 @@
+    This function is idempotent and safe to call multiple times. It checks if
+    handlers are already configured to avoid duplication.
+    """
+    root_logger = logging.getLogger()
+    if not root_logger.hasHandlers():
+        root_logger.setLevel(level)
+        handler = logging.StreamHandler(sys.stdout)
+        handler.setFormatter(JsonFormatter())
+        root_logger.addHandler(handler)
```
How about something like that?
```
    logging.basicConfig(
        # filename='logs/log_file_name.log',
        level=logging.INFO,
        format='[%(asctime)s] %(filename)-15s:%(lineno)-4d %(levelname)-7s - %(message)s',
        datefmt='%H:%M:%S'
    )
```

The version I wrote with the commented filename goes to the stdout, otherwise it writes a log file into the folder `rre-dataset-generator/logs` (in that case, we need to create that logs folder to avoid exceptions.

dantuzi replied:
Bear in mind that if you uncomment the filename, it works correctly only when you run the script from the `dataset_generator.py` entry point. If you want to test the code below, it will fail

dantuzi commented on rre-dataset-generator/tests/unit/test_dataset_generation_dummy.py:
```diff
@@ -0,0 +1,40 @@
+def args():
+    return parse_args()
+
+# def test_argparse_yaml_with_list(caplog):
+#     # Set up logging capture
+#     caplog.set_level(logging.INFO)
+#
+#     runner = CliRunner()
+#     # Pass config_path as a positional argument
+#     result = runner.invoke(app, ["src/yaml_configs/config.yml"])
+#
+#     # Print the logs for debugging
+#     for record in caplog.records:
+#         print(f"LOG: {record.message}")
+#
+#     # Check if the log message appears in any of the log records
+#     assert any("Starting dataset generation process..." in record.message
+#               for record in caplog.records)
+#     assert result.exit_code == 0
```
I would avoid committing the commented code like that

dantuzi commented on rre-dataset-generator/README.md:
````diff
@@ -1,2 +1,109 @@
+
+Now that the environment is active, execute `pytest` command as follows:
+```bash
+pytest tests/
````
from the bash it works only the command `pytest`

dantuzi commented on rre-dataset-generator/uv.lock:
Do we need to commit this file?

dantuzi commented on rre-dataset-generator/.gitignore:
```diff
@@ -0,0 +1,14 @@
+
+# Local Files and Directories
+.output_dataset
+output/
```
Don't we need to add the `uv.lock` file?

nicolo-rinaldi committed:
Addressed comments [PR#150](https://github.com/SeaseLtd/rated-ranking-evaluator/pull/150/files/d3a000e9cf6ae284ac03a469e53ff5d38ee9cc4d)

## Changes
- removed file uv.lock and add it in .gitignore
- simplified logger.py
- added logs/ folder
- removed Makefile
- resolved conflicts in pyproject.toml
- simplified test file test_dataset_generation_dummy.py
